quickwit/quickwit-indexing/src/source/queue_sources/processor.rs

```diff
@@ -0,0 +1,464 @@
+            storage_resolver: source_runtime.storage_resolver,
+            queue,
+            observable_state: QueueProcessorObservableState::default(),
+            queue_params,
+            publish_lock: PublishLock::default(),
+            publish_token: Ulid::new().to_string(),
+        }
+    }
+
+    pub async fn initialize(
+        &mut self,
+        doc_processor_mailbox: &Mailbox<DocProcessor>,
+        ctx: &SourceContext,
+    ) -> Result<(), ActorExitStatus> {
+        let publish_lock = self.publish_lock.clone();
+        ctx.send_message(doc_processor_mailbox, NewPublishLock(publish_lock))
+            .await?;
+        ctx.send_message(
+            doc_processor_mailbox,
+            NewPublishToken(self.publish_token.clone()),
+        )
+        .await?;
+        Ok(())
+    }
+
+    /// Starts background tasks that extend the visibility deadline of the
+    /// messages until they are dropped.
+    fn spawn_visibility_tasks(
+        &mut self,
+        messages: Vec<CheckpointedMessage>,
+    ) -> Vec<(CheckpointedMessage, VisibilityTaskHandle)> {
+        messages
+            .into_iter()
+            .map(|message| {
+                let handle = spawn_visibility_task(
+                    self.queue.clone(),
+                    message.content.metadata.ack_id.clone(),
+                    message.content.metadata.initial_deadline,
+                    self.publish_lock.clone(),
+                );
+                (message, handle)
+            })
+            .collect()
+    }
+
+    /// Polls messages from the queue and prepares them for processing
+    async fn poll_messages(&mut self, ctx: &SourceContext) -> anyhow::Result<()> {
+        // receive() typically uses long polling so it can be long to respond
+        let raw_messages = ctx.protect_future(self.queue.receive()).await?;
+
+        if raw_messages.is_empty() {
+            self.observable_state.num_consecutive_empty_batches += 1;
+            return Ok(());
+        } else {
+            self.observable_state.num_consecutive_empty_batches = 0;
+        }
+
+        let preprocessed_messages = raw_messages
+            .into_iter()
+            .map(|m| m.pre_process(self.queue_params.message_type))
+            .collect::<anyhow::Result<Vec<_>>>()?;
+
+        let categorized_using_local_state =
+            self.local_state.filter_completed(preprocessed_messages);
+
+        let categorized_using_shared_state = self
+            .shared_state
+            .checkpoint_messages(
+                &self.publish_token,
+                categorized_using_local_state.processable,
+            )
+            .await?;
+
+        // Drop visibility tasks for messages that have been processed by another pipeline
+        let completed_visibility_tasks = categorized_using_shared_state
+            .already_processed
+            .iter()
+            .filter_map(|m| self.local_state.mark_completed(m.partition_id()))
```

Comment:
```suggestion
            .filter_map(|m| self.local_state.mark_completed(m.partition_id()))
```

please avoid one letter variable.